src/compiler/core.ts

```diff
@@ -655,24 +659,85 @@ namespace ts {
         return [...array1, ...array2];
     }
 
-    // TODO: fixme (N^2) - add optional comparer so collection can be sorted before deduplication.
-    export function deduplicate<T>(array: ReadonlyArray<T>, areEqual?: (a: T, b: T) => boolean): T[] {
-        let result: T[];
-        if (array) {
-            result = [];
-            loop: for (const item of array) {
-                for (const res of result) {
-                    if (areEqual ? areEqual(res, item) : res === item) {
-                        continue loop;
-                    }
-                }
-                result.push(item);
+    function deduplicateRelational<T>(array: ReadonlyArray<T>, equalityComparer: EqualityComparer<T>, comparer: Comparer<T>) {
+        // Perform a stable sort of the array. This ensures the first entry in a list of
+        // duplicates remains the first entry in the result.
+        const indices = array.map((_, i) => i);
+        stableSortIndices(array, indices, comparer);
+
+        let last = array[indices[0]];
+        const deduplicated: number[] = [indices[0]];
+        for (let i = 1; i < indices.length; i++) {
+            const index = indices[i];
+            const item = array[index];
+            if (!equalityComparer(last, item)) {
+                deduplicated.push(index);
+                last = item;
             }
         }
+
+        // restore original order
+        deduplicated.sort();
+        return deduplicated.map(i => array[i]);
+    }
+
+    function deduplicateEquality<T>(array: ReadonlyArray<T>, equalityComparer: EqualityComparer<T>) {
+        const result: T[] = [];
+        for (const item of array) {
+            pushIfUnique(result, item, equalityComparer);
+        }
         return result;
     }
 
-    export function arrayIsEqualTo<T>(array1: ReadonlyArray<T>, array2: ReadonlyArray<T>, equaler?: (a: T, b: T) => boolean): boolean {
+    /**
+     * Deduplicates an unsorted array.
+     * @param equalityComparer An optional `EqualityComparer` used to determine if two values are duplicates.
+     * @param comparer An optional `Comparer` used to sort entries before comparison. If supplied,

```

Comment:
If a comparer isn't supplied then we should also keep the original order because we push in that order.

Reply:
Results are always supplied in the original order. I'll update the comment.